lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+}
+
+typedef Ufe::Transform3d::Ptr (*NextTransform3dFn)(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item
+);
+
+Ufe::Transform3d::Ptr createTransform3d(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item,
+    NextTransform3dFn                   nextTransform3dFn
+)
+{
+    UsdSceneItem::Ptr usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+#if !defined(NDEBUG)
+    assert(usdItem);
+#endif
+
+    // If the prim isn't transformable, can't create a Transform3d interface
+    // for it.
+    UsdGeomXformable xformSchema(usdItem->prim());
+    if (!xformSchema) {
+        return nullptr;
+    }
+    bool resetsXformStack = false;
+    auto xformOps = xformSchema.GetOrderedXformOps(&resetsXformStack);
+
+    // Early out: if there are no transform ops yet, it's a match.
+    if (xformOps.empty()) {
+        return UsdTransform3dMayaXformStack::create(usdItem);
+    }
+
+    // If the prim supports the Maya transform stack, create a Maya transform
+    // stack interface for it, otherwise delegate to the next handler in the
+    // chain of responsibility.
+    auto stackOps = UsdMayaXformStack::MayaStack().MatchingSubstack(xformOps);
+
+    return stackOps.empty() ? nextTransform3dFn(nextHandler, item) :
+        UsdTransform3dMayaXformStack::create(usdItem);
+}
+
+// Helper class to factor out common code for translate, rotate, scale
+// undoable commands.
+class UsdTRSUndoableCmdBase : public Ufe::SetVector3dUndoableCommand {
+private:
+    UsdGeomXformOp    fOp;
+    const UsdTimeCode fReadTime;
+    const UsdTimeCode fWriteTime;
+    const VtValue     fPrevOpValue;
+    const TfToken     fAttrName;
+
+public:
+    UsdTRSUndoableCmdBase(
+        const VtValue&        newOpValue,
+        const Ufe::Path&      path,
+        const UsdGeomXformOp& op,
+        const UsdTimeCode&    writeTime_
+    ) : Ufe::SetVector3dUndoableCommand(path), fOp(op),
+        // Always read from proxy shape time.
+        fReadTime(getTime(path)),
+        fWriteTime(writeTime_),
+        fPrevOpValue(getValue(op.GetAttr(), readTime())),
+        fNewOpValue(newOpValue),
+        fAttrName(op.GetOpName())
+    {}
+
+    // Have separate execute override which is value-only, as default execute()
+    // calls redo, and undo / redo will eventually not be value only and will
+    // support propertySpec / primSpec management.
+    void execute() override { setValue(fNewOpValue); }
+
+    void recreateOp() {
+        auto sceneItem = std::dynamic_pointer_cast<UsdSceneItem>(
+            Ufe::Hierarchy::createItem(path())); TF_AXIOM(sceneItem);
+        auto prim = sceneItem->prim(); TF_AXIOM(prim);
+        auto attr = prim.GetAttribute(fAttrName); TF_AXIOM(attr);
+        fOp       = UsdGeomXformOp(attr);
+    }
+
+    void undo() override {
+        // Re-create the XformOp, as the underlying prim may have been
+        // invalidated by later commands.
```

Comment:
Our implementation of parenting operations actually removes prims, so if there's a parenting command in the undo stack either before or after us, we need to re-create the transform op.